File: python/canvastekk_workflow_sdk/response.py
"""
Node Execution Response Model

This is the response returned from a node's /execute endpoint.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class NodeExecutionResponse(BaseModel):
    """
    Response payload from node execution.

    Returned by a node's POST /execute endpoint (sync) or
    POSTed to callback_url (async).
    """

    execution_id: str = Field(description="Unique ID for this execution (for cancellation)")
    status: Literal["pass", "fail"] = Field(description="Execution result status")
    outputs: dict[str, Any] | None = Field(
        default=None,
        description="Output values (if pass)",
    )
    token_usage: float = Field(
        default=0.0,
        ge=0.0,
        description="Actual tokens consumed",
    )
    duration_ms: int = Field(
        default=0,
        ge=0,
        description="Execution time in milliseconds",
    )
    error: str | None = Field(
        default=None,
        description="Error message (if fail)",
    )
    error_type: str | None = Field(
        default=None,
        description="Error class name (if fail)",
    )
    error_code: str | None = Field(
        default=None,
        description="Structured error code (e.g., TIMEOUT, VALIDATION_ERROR)",
    )

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "execution_id": "exec-xyz789",
                    "status": "pass",
                    "outputs": {"message": "Hello, World!", "timestamp": "2026-01-15T10:30:00+00:00"},
                    "token_usage": 0.0,
                    "duration_ms": 42,
                    "error": None,
                    "error_type": None,
                    "error_code": None,
                }
            ]
        }
    }

    @classmethod
    def success(
        cls,
        execution_id: str,
        outputs: dict[str, Any],
        duration_ms: int = 0,
        token_usage: float = 0.0,
    ) -> NodeExecutionResponse:
        """Create a successful execution response.

        Args:
            execution_id: Unique identifier for this execution.
            outputs: Output values matching the node's output_schema.
            duration_ms: Execution time in milliseconds.
            token_usage: Tokens consumed during execution.

        Returns:
            A :class:`NodeExecutionResponse` with ``status="pass"``.
        """
        return cls(
            execution_id=execution_id,
            status="pass",
            outputs=outputs,
            duration_ms=duration_ms,
            token_usage=token_usage,
        )

    @classmethod
    def failure(
        cls,
        execution_id: str,
        error: str,
        error_type: str | None = None,
        duration_ms: int = 0,
        error_code: str | None = None,
    ) -> NodeExecutionResponse:
        """Create a failed execution response.

        Args:
            execution_id: Unique identifier for this execution.
            error: Human-readable error message.
            error_type: Exception class name (e.g. ``"NodeTimeoutError"``).
            duration_ms: Execution time before failure.
            error_code: Structured error code (e.g. ``"TIMEOUT"``).

        Returns:
            A :class:`NodeExecutionResponse` with ``status="fail"``.
        """
        return cls(
            execution_id=execution_id,
            status="fail",
            outputs=None,
            error=error,
            error_type=error_type,
            duration_ms=duration_ms,
            error_code=error_code,
        )


class HealthResponse(BaseModel):
    """Response from /health endpoint."""

    status: Literal["healthy", "unhealthy", "degraded"] = Field(
        description="Health status",
    )
    node_id: str = Field(description="Node identifier with version")
    version: str = Field(description="Node version")
    checks: dict[str, bool] = Field(
        default_factory=dict,
        description="Individual health check results",
    )

    model_config = {
        "json_schema_extra": {
            "examples": [
                {
                    "status": "healthy",
                    "node_id": "echo-v1.0.0",
                    "version": "1.0.0",
                    "checks": {},
                }
            ]
        }
    }

File: typescript/src/response.ts
import { z } from "zod";

export const NodeExecutionResponseSchema = z.object({
  execution_id: z.string(),
  status: z.union([z.literal("pass"), z.literal("fail")]),
  outputs: z.record(z.unknown()).nullable().default(null),
  token_usage: z.number().min(0.0).default(0.0),
  duration_ms: z.number().int().min(0).default(0),
  error: z.string().nullable().default(null),
  error_type: z.string().nullable().default(null),
  error_code: z.string().nullable().default(null),
});

/**
 * Node execution response returned to the workflow engine.
 *
 * Fields use snake_case for wire-format compatibility with the Python engine.
 *
 * @property execution_id - Unique identifier for this execution
 * @property status - Execution result: `"pass"` or `"fail"`
 * @property outputs - Output values on success, `null` on failure
 * @property duration_ms - Execution duration in milliseconds
 * @property token_usage - Token consumption (0.0 if not applicable)
 * @property error - Error message on failure, `null` on success
 * @property error_type - Error class name (e.g., `"NodeValidationError"`)
 * @property error_code - Machine-readable error code (e.g., `"VALIDATION_ERROR"`)
 */
export type NodeExecutionResponse = z.infer<typeof NodeExecutionResponseSchema>;

/**
 * Creates a success response from node execution.
 *
 * @param executionId - Unique execution identifier
 * @param outputs - Node output values
 * @param durationMs - Execution duration in milliseconds
 * @param tokenUsage - Token consumption count
 * @returns NodeExecutionResponse with status "pass"
 */
function createSuccessResponse(
  executionId: string,
  outputs: Record<string, unknown>,
  durationMs = 0,
  tokenUsage = 0.0,
): NodeExecutionResponse {
  return {
    execution_id: executionId,
    status: "pass",
    outputs,
    duration_ms: durationMs,
    token_usage: tokenUsage,
    error: null,
    error_type: null,
    error_code: null,
  };
}

/**
 * Creates a failure response from node execution.
 *
 * @param executionId - Unique execution identifier
 * @param error - Human-readable error message
 * @param errorType - Error class name
 * @param durationMs - Execution duration in milliseconds
 * @param errorCode - Machine-readable error code
 * @returns NodeExecutionResponse with status "fail"
 */
function createFailureResponse(
  executionId: string,
  error: string,
  errorType: string | null = null,
  durationMs = 0,
  errorCode: string | null = null,
): NodeExecutionResponse {
  return {
    execution_id: executionId,
    status: "fail",
    outputs: null,
    error,
    error_type: errorType,
    duration_ms: durationMs,
    error_code: errorCode,
    token_usage: 0.0,
  };
}

export const NodeExecutionResponseFactory = {
  success: createSuccessResponse,
  failure: createFailureResponse,
};

export const HealthResponseSchema = z.object({
  status: z.union([
    z.literal("healthy"),
    z.literal("unhealthy"),
    z.literal("degraded"),
  ]),
  node_id: z.string(),
  version: z.string(),
  checks: z.record(z.boolean()).default({}),
});

/** Node health check response. */
export type HealthResponse = z.infer<typeof HealthResponseSchema>;
